File: the knight tour.php
<?php
define("N", 8);

// Possible moves of a Knight in chess
$moveX = [2, 1, -1, -2, -2, -1, 1, 2];
$moveY = [1, 2, 2, 1, -1, -2, -2, -1];

// Function to print the board
function printSolution($board) {
    for ($x = 0; $x < N; $x++) {
        for ($y = 0; $y < N; $y++) {
            printf("%3d ", $board[$x][$y]);
        }
        echo "\n";
    }
}

// Function to check if a move is valid
function isSafe($x, $y, $board) {
    return ($x >= 0 && $x < N && $y >= 0 && $y < N && $board[$x][$y] == -1);
}

// Recursive utility function to solve Knight’s Tour
function solveKTUtil($x, $y, $movei, &$board, $moveX, $moveY) {
    if ($movei == N * N)
        return true; // All squares are visited

    for ($k = 0; $k < 8; $k++) {
        $next_x = $x + $moveX[$k];
        $next_y = $y + $moveY[$k];
        if (isSafe($next_x, $next_y, $board)) {
            $board[$next_x][$next_y] = $movei;
            if (solveKTUtil($next_x, $next_y, $movei + 1, $board, $moveX, $moveY))
                return true;
            else
                $board[$next_x][$next_y] = -1; // backtrack
        }
    }

    return false;
}

// Main solver function
function solveKT($startX, $startY) {
    global $moveX, $moveY;

    // Initialize the board with -1
    $board = array_fill(0, N, array_fill(0, N, -1));

    // Start from user’s input
    $board[$startX][$startY] = 0;

    if (!solveKTUtil($startX, $startY, 1, $board, $moveX, $moveY)) {
        echo "Solution does not exist.\n";
        return;
    }

    printSolution($board);
}

// --- Program execution ---

// You can change the starting position here
$startX = 0;
$startY = 0;

// If run via CLI, optionally get coordinates from input
if (PHP_SAPI === 'cli' && isset($argv[1]) && isset($argv[2])) {
    $startX = intval($argv[1]);
    $startY = intval($argv[2]);
} else {
    echo "Using default starting position (0,0)\n";
}

solveKT($startX, $startY);
?>

File: the knight tour.cpp
#include <iostream>
#include <iomanip>
using namespace std;

#define N 8 // You can change this to any board size, e.g., 5, 6, 8, 10

// Moves a knight can make
int moveX[8] = {2, 1, -1, -2, -2, -1, 1, 2};
int moveY[8] = {1, 2, 2, 1, -1, -2, -2, -1};

// Function to check if a position is valid and not yet visited
bool isSafe(int x, int y, int sol[N][N]) {
    return (x >= 0 && x < N && y >= 0 && y < N && sol[x][y] == -1);
}

// Function to print the solution board
void printSolution(int sol[N][N]) {
    for (int x = 0; x < N; x++) {
        for (int y = 0; y < N; y++)
            cout << setw(3) << sol[x][y] << " ";
        cout << endl;
    }
}

// Recursive utility function to solve Knight’s Tour
bool solveKTUtil(int x, int y, int movei, int sol[N][N], int moveX[N], int moveY[N]) {
    int nextX, nextY;
    if (movei == N * N)
        return true; // All squares visited

    // Try all possible moves
    for (int k = 0; k < 8; k++) {
        nextX = x + moveX[k];
        nextY = y + moveY[k];
        if (isSafe(nextX, nextY, sol)) {
            sol[nextX][nextY] = movei;
            if (solveKTUtil(nextX, nextY, movei + 1, sol, moveX, moveY))
                return true;
            else
                sol[nextX][nextY] = -1; // backtracking
        }
    }

    return false;
}

// Main function to solve the Knight’s Tour problem
bool solveKT() {
    int sol[N][N];

    // Initialize the solution matrix
    for (int x = 0; x < N; x++)
        for (int y = 0; y < N; y++)
            sol[x][y] = -1;

    // Knight’s starting position
    int startX, startY;
    cout << "Enter starting position (x y): ";
    cin >> startX >> startY;

    // Place the knight on the starting square
    sol[startX][startY] = 0;

    // Try to solve the Knight’s Tour
    if (!solveKTUtil(startX, startY, 1, sol, moveX, moveY)) {
        cout << "No solution exists." << endl;
        return false;
    } else
        printSolution(sol);

    return true;
}

int main() {
    cout << "=== Knight's Tour Problem (C++) ===" << endl;
    solveKT();
    return 0;
}
